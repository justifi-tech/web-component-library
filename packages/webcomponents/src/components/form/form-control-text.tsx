import {
  Component,
  Host,
  h,
  Prop,
  Event,
  EventEmitter,
  State,
  Element,
  Watch,
} from '@stencil/core';

@Component({
  tag: 'form-control-text'
})
export class TextInput {
  @Element() el: HTMLElement;

  @Prop() label: string;
  @Prop() name: any;
  @Prop() helpText?: string;
  @Prop() errorText?: string;
  @Prop() defaultValue: string;
  @Prop() maxLength?: number;
  @Prop() inputHandler: (name: string, value: string) => void;
  @Prop() keyDownHandler?: (event: any) => void;
  @Prop() disabled: boolean;

  @State() input: string;

  @Event() formControlInput: EventEmitter<any>;
  @Event() formControlBlur: EventEmitter<any>;

  @Watch('defaultValue')
  handleDefaultValueChange(newValue: string) {
    this.updateInput(newValue);
  }

  updateInput(newValue: any) {
    const inputElement = this.el.querySelector('input');
    if (inputElement) {
      inputElement.value = newValue || '';
    }
  }

  handleFormControlInput(event: any) {
    const target = event.target;
    const name = target.getAttribute('name');
    this.inputHandler(name, target.value);
    this.formControlInput.emit(target.value);
  }

  componentDidLoad() {
    this.updateInput(this.defaultValue);
  }

  render() {
    return (
      <Host exportparts="label,input,input-invalid">
        <label part="label" class="form-label" htmlFor={this.name}>
          {this.label}
        </label>
        <input
          id={this.name}
          name={this.name}
          onInput={(event: any) => this.handleFormControlInput(event)}
          onBlur={() => this.formControlBlur.emit()}
          onKeyDown={(event: any) => this.keyDownHandler && this.keyDownHandler(event)}
          maxLength={this.maxLength}
          part={`input ${this.errorText ? 'input-invalid ' : ''}${this.disabled ? ' input-disabled' : ''}`}
          class={this.errorText ? 'form-control is-invalid' : 'form-control'}
          type="text"
          disabled={this.disabled}
        />
        <form-control-help-text helpText={this.helpText} />
        <br />
        <form-control-error-text errorText={this.errorText} />
      </Host>
    );
  }
}

