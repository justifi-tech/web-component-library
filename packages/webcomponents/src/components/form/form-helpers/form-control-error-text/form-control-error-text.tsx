import { Component, Prop, h } from '@stencil/core';

@Component({
  tag: 'form-control-error-text',
})
export class FormControlErrorText {
  @Prop() errorText: string;
  @Prop() name: string;

  render() {

    if (!this.errorText) {
      return null;
    }

    return (
      <small class='form-text text-danger' id={`form-error-text-${this.name}`}>
        {this.errorText}
      </small>
    );
  }
}
