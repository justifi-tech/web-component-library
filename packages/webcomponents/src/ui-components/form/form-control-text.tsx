import {
  Component,
  Host,
  h,
  Prop,
  Event,
  EventEmitter,
  Element,
  Watch,
} from '@stencil/core';

@Component({
  tag: 'form-control-text'
})
export class TextInput {
  @Element() el: HTMLElement;

  @Prop() name: any;
  @Prop() label: string;
  @Prop() inputHandler: (name: string, value: string) => void;
  @Prop() defaultValue?: string;
  @Prop() helpText?: string;
  @Prop() errorText?: string;
  @Prop() maxLength?: number;
  @Prop() keyDownHandler?: (event: any) => void;
  @Prop() disabled?: boolean;

  @Watch('defaultValue')
  handleDefaultValueChange(newValue: string) {
    this.updateInput(newValue);
  }

  @Event() formControlInput: EventEmitter<any>;
  @Event() formControlBlur: EventEmitter<any>;

  componentDidLoad() {
    this.updateInput(this.defaultValue);
  }

  handleFormControlInput = (event: any) => {
    const target = event.target;
    const name = target.getAttribute('name');
    this.inputHandler(name, target.value);
    this.formControlInput.emit({ name, value: target.value });
  }

  updateInput = (newValue: any) => {
    const inputElement = this.el.querySelector('input');
    if (inputElement) {
      inputElement.value = newValue || '';
    }
  }

  render() {
    return (
      <Host exportparts="label,input,input-invalid">
        <div class="form-group d-flex flex-column">
          <div class="d-flex align-items-start gap-2">
            <label part="label" class="form-label" htmlFor={this.name}>
              {this.label}
            </label>
            <form-control-tooltip helpText={this.helpText} />
          </div>
          <input
            id={this.name}
            name={this.name}
            onBlur={this.formControlBlur.emit}
            onInput={this.handleFormControlInput}
            onKeyDown={this.keyDownHandler}
            onPaste={this.keyDownHandler}
            maxLength={this.maxLength}
            part={`input ${this.errorText ? "input-invalid " : ""}${this.disabled ? " input-disabled" : ""}`}
            class={this.errorText ? "form-control is-invalid" : "form-control"}
            type="text"
            disabled={this.disabled}
          />
          <form-control-error-text errorText={this.errorText} name={this.name} />
        </div>
      </Host>
    );
  }
}
