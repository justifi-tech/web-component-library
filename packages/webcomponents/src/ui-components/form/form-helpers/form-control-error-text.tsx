import { h } from '@stencil/core';

export const FormControlErrorText = (props: { errorText: string, name: string }) => {
  const { errorText, name } = props;

  if (!errorText) {
    return null;
  }

  return (
    <small class='form-text text-danger' id={`form-error-text-${name}`} part="input-error-text" >
      {errorText}
    </small>
  );
}
