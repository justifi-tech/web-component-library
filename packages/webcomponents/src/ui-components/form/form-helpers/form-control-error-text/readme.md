# form-control-error-text



<!-- Auto Generated Below -->


## Properties

| Property    | Attribute    | Description | Type     | Default     |
| ----------- | ------------ | ----------- | -------- | ----------- |
| `errorText` | `error-text` |             | `string` | `undefined` |
| `name`      | `name`       |             | `string` | `undefined` |


## Dependencies

### Used by

 - [form-control-checkbox](../..)
 - [form-control-date](../..)
 - [form-control-file](../..)
 - [form-control-file-v2](../..)
 - [form-control-monetary](../..)
 - [form-control-number](../..)
 - [form-control-number-masked](../..)
 - [form-control-select](../..)
 - [form-control-text](../..)
 - [form-control-textarea](../..)
 - [justifi-dispute-reason](../../../../components/dispute-management/dispute-response)

### Graph
```mermaid
graph TD;
  form-control-checkbox --> form-control-error-text
  form-control-date --> form-control-error-text
  form-control-file --> form-control-error-text
  form-control-file-v2 --> form-control-error-text
  form-control-monetary --> form-control-error-text
  form-control-number --> form-control-error-text
  form-control-number-masked --> form-control-error-text
  form-control-select --> form-control-error-text
  form-control-text --> form-control-error-text
  form-control-textarea --> form-control-error-text
  justifi-dispute-reason --> form-control-error-text
  style form-control-error-text fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
