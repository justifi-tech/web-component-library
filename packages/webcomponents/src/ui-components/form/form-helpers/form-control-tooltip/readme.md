# form-control-tooltip



<!-- Auto Generated Below -->


## Properties

| Property   | Attribute   | Description | Type     | Default     |
| ---------- | ----------- | ----------- | -------- | ----------- |
| `helpText` | `help-text` |             | `string` | `undefined` |


## Dependencies

### Used by

 - [form-control-date](../..)
 - [form-control-file](../..)
 - [form-control-file-v2](../..)
 - [form-control-number-masked](../..)
 - [form-control-select](../..)
 - [form-control-text](../..)
 - [justifi-additional-questions-form-step-core](../../../../components/business-forms/payment-provisioning/additional-questions)
 - [justifi-business-bank-account-form-step-core](../../../../components/business-forms/payment-provisioning/bank-account)
 - [justifi-business-owners-form-step-core](../../../../components/business-forms/payment-provisioning/business-owners)
 - [justifi-business-representative-form-inputs](../../../../components/business-forms/payment-provisioning/business-representative)
 - [justifi-legal-address-form-step-core](../../../../components/business-forms/payment-provisioning/legal-address-form)
 - [terminal-quantity-selector](../../../../components/order-terminals/terminal-quantity-selector)

### Depends on

- [custom-popper](../../../custom-popper)

### Graph
```mermaid
graph TD;
  form-control-tooltip --> custom-popper
  form-control-date --> form-control-tooltip
  form-control-file --> form-control-tooltip
  form-control-file-v2 --> form-control-tooltip
  form-control-number-masked --> form-control-tooltip
  form-control-select --> form-control-tooltip
  form-control-text --> form-control-tooltip
  justifi-additional-questions-form-step-core --> form-control-tooltip
  justifi-business-bank-account-form-step-core --> form-control-tooltip
  justifi-business-owners-form-step-core --> form-control-tooltip
  justifi-business-representative-form-inputs --> form-control-tooltip
  justifi-legal-address-form-step-core --> form-control-tooltip
  terminal-quantity-selector --> form-control-tooltip
  style form-control-tooltip fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
