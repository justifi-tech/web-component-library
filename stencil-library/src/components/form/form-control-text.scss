@import 'reboot';
@import 'forms';

:host {
  @include host-base-styles;
  display: block;
}

input[type='text'][disabled] {
  color: #495057;
  background-color: #e9ecef;
  opacity: 0.5;
  cursor: not-allowed;
}
