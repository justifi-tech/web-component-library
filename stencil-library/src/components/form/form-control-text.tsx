import { Component, Host, h, Prop } from '@stencil/core';

@Component({
  tag: 'form-control-text',
  styleUrl: 'form-control-text.scss',
  shadow: true,
})
export class TextInput {
  @Prop() label: string;
  @Prop() name: any;
  @Prop() onInput: (e: any) => void;
  @Prop() error: string;
  @Prop() value: string;

  render() {
    return (
      <Host exportparts="label,input,input-invalid">
        <label part="label" class="form-label">
          {this.label}
        </label>
        <input
          name={this.name}
          type="text"
          onInput={this.onInput}
          value={this.value}
          part={`input ${this.error && 'input-invalid'}`}
          class={this.error ? 'form-control is-invalid' : 'form-control'}
        />
        {this.error && <div class="invalid-feedback">{this.error}</div>}
      </Host>
    );
  }
}
