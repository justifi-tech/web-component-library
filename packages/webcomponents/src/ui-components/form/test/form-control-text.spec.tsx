import { h } from '@stencil/core';
import { newSpecPage } from '@stencil/core/testing';
import { TextInput } from '../form-control-text';
import { FormControlErrorText } from '../form-helpers/form-control-error-text/form-control-error-text';
import { TooltipComponent } from '../form-helpers/form-control-tooltip/form-control-tooltip';

describe('form-control-text', () => {
  const components = [TextInput, FormControlErrorText, TooltipComponent];
  const mockInputHandler = jest.fn();
  const mockKeyDownHandler = jest.fn();

  it('Renders with default props', async () => {
    const page = await newSpecPage({
      components: components,
      template: () => <form-control-text label='Username' name='username'/>,
    });

    expect(page.root).toMatchSnapshot();
    expect(page.rootInstance.label).toBe('Username');
    expect(page.rootInstance.name).toBe('username');
  });

  it('Renders with all props provided', async () => {
    const page = await newSpecPage({
      components: components,
      template: () =>
        <form-control-text
          label='Email'
          name='email'
          defaultValue='user@example.com'
          errorText='Invalid email'
          helpText='Enter your email'
          disabled
          maxLength={50}
          inputHandler={mockInputHandler}
          keyDownHandler={mockKeyDownHandler}
        />
    });

    expect(page.root).toMatchSnapshot();
  });

  it('Updates input value on defaultValue prop change', async () => {
    // Initial render with the default value
    let defaultValue = 'user@example.com';
    let page = await newSpecPage({
      components: components,
      template: () =>
        <form-control-text
          label='Email'
          name='email'
          defaultValue={defaultValue}
        />
    });

    let inputElement = page.root.querySelector('input');
    expect(inputElement.value).toBe('user@example.com');

    defaultValue = 'updated@example.com';

    // Re-render component with new defaultValue
    page = await newSpecPage({
      components: components,
      template: () =>
        <form-control-text
          label='Email'
          name='email'
          defaultValue={defaultValue}
        />
    });

    inputElement = page.root.querySelector('input');
    expect(inputElement.value).toBe('updated@example.com');
  });

  it('Handles user input correctly', async () => {
    const page = await newSpecPage({
      components: components,
      template: () =>
        <form-control-text
          label='Email'
          name='email'
          defaultValue='user@example.com'
          inputHandler={mockInputHandler}
        />
    });

    const inputElement = page.root.querySelector('input');
    const testValue = 'Hello, World!';

    inputElement.value = testValue;
    await inputElement.dispatchEvent(new Event('input'));

    expect(inputElement.value).toBe(testValue);
  });

  it('Emits formControlInput event on input', async () => {
    const page = await newSpecPage({
      components: components,
      template: () =>
        <form-control-text
          label='Email'
          name='email'
          defaultValue='user@example.com'
          inputHandler={mockInputHandler}
        />
    });

    const inputEventSpy = jest.fn();
    page.root.addEventListener('formControlInput', inputEventSpy);

    await page.waitForChanges();

    const inputElement = page.root.querySelector('input');
    inputElement.value = 'other-user@example.com';

    inputElement.dispatchEvent(new Event('input', { bubbles: true, composed: true }));

    await page.waitForChanges();

    expect(inputEventSpy).toHaveBeenCalledWith(expect.objectContaining({
      detail: {
        name: 'email',
        value: 'other-user@example.com'
      }
    }));
  });

  it('Emits formControlBlur event on blur', async () => {
    const page = await newSpecPage({
      components: components,
      template: () => <form-control-text defaultValue='user@example.com' inputHandler={mockInputHandler}/>
    });

    const blurEventSpy = jest.fn();
    page.root.addEventListener('formControlBlur', blurEventSpy);

    const inputElement = page.root.querySelector('input');
    inputElement.dispatchEvent(new Event('blur'));

    expect(blurEventSpy).toHaveBeenCalled();
  });

  it('Disables input when disabled prop is provided', async () => {
    const page = await newSpecPage({
      components: components,
      template: () => <form-control-text label='Username' name='username' disabled/>,
    });

    const inputElement = page.root.querySelector('input');
    expect(inputElement.disabled).toBeTruthy();
  });

  it('Shows help text when helpText prop is provided', async () => {
    const page = await newSpecPage({
      components: components,
      template: () => <form-control-text label='Username' name='username' helpText='This is the help text.'/>,
    });

    const tooltipComponent = page.root.querySelector('form-control-tooltip');
    expect(tooltipComponent).not.toBeNull();

    const tooltipIcon = tooltipComponent.querySelector('.bi-question-square');
    expect(tooltipIcon).not.toBeNull();

    const tooltipElement = tooltipComponent.querySelector('.tooltip');
    expect(tooltipElement).not.toBeNull();

    const tooltipText = tooltipElement.querySelector('.tooltip-inner');
    expect(tooltipText.textContent).toBe('This is the help text.');
  });

  it('Shows error and applies error styling when error prop is provided', async () => {
    const page = await newSpecPage({
      components: components,
      template: () => <form-control-text label='Username' name='username' errorText='This field is required.'/>,
    });

    const errorTextComponent = page.root.querySelector('form-control-error-text');
    expect(errorTextComponent).not.toBeNull();

    const errorText = errorTextComponent.querySelector('.text-danger');
    expect(errorText.textContent).toBe('This field is required.');

    const inputElement = page.root.querySelector('input');
    expect(inputElement.classList.contains('is-invalid')).toBe(true);
  });
});
