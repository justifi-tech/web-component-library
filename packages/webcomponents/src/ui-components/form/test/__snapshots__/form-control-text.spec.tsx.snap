// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`form-control-text Renders with all props provided 1`] = `
<form-control-text exportparts="label,input,input-invalid">
  <div class="d-flex flex-column form-group">
    <label class="form-label" htmlfor="email" part="label">
      Email
    </label>
    <input class="form-control is-invalid" disabled="" id="email" maxlength="50" name="email" part="input input-invalid  input-disabled" type="text" value="user@example.com">
    <form-control-help-text>
      <small class="form-text text-muted" id="form-help-text-email">
        Enter your email
      </small>
    </form-control-help-text>
    <form-control-error-text>
      <small class="form-text text-danger" id="form-error-text-email">
        Invalid email
      </small>
    </form-control-error-text>
  </div>
</form-control-text>
`;

exports[`form-control-text Renders with default props 1`] = `
<form-control-text exportparts="label,input,input-invalid">
  <div class="d-flex flex-column form-group">
    <label class="form-label" htmlfor="username" part="label">
      Username
    </label>
    <input class="form-control" id="username" name="username" part="input " type="text" value="">
    <form-control-help-text></form-control-help-text>
    <form-control-error-text></form-control-error-text>
  </div>
</form-control-text>
`;
